credential.go

```diff
@@ -0,0 +1,84 @@
+package mls
+
+import (
+	"fmt"
+	"github.com/bifurcation/mint/syntax"
+)
+
+type CredentialType uint8
+
+const (
+	credential_basic = iota
+	credential_x509
+)
+
+// enum {
+//       ecdsa_secp256r1_sha256(0x0403),
+//       ed25519(0x0807),
+//       (0xFFFF)
+//   } SignatureScheme;
+type SignatureScheme uint16
+
+// struct {
+//     opaque identity<0..2^16-1>;
+//     SignatureScheme algorithm;
+//     SignaturePublicKey public_key;
+// } BasicCredential;
+type BasicCredential struct {
+	Identity           []byte `tls:"head=2"`
+	SignatureScheme    SignatureScheme
+	SignaturePublicKey []byte `tls:"head=2"`
+}
+
+type X509Credential struct {
+	CertData []byte `tls:"head=3"`
+}
+
+//	struct {
+//		CredentialType credential_type;
+//		select (Credential.credential_type) {
+//			case basic:
+//				BasicCredential;
+//			case x509:
+//				opaque cert_data<1..2^24-1>;
+//		};
+//} Credential;
+type Credential struct {
+	CredentialType  CredentialType
+	BasicCredential BasicCredential
+	X509Credential  X509Credential
+}
+
+func (c Credential) MarshalTLS() ([]byte, error) {
+	switch c.CredentialType {
+	case credential_basic:
+		return syntax.Marshal(c.BasicCredential)
+	case credential_x509:
+		return syntax.Marshal(c.X509Credential)
+	default:
+		return nil, fmt.Errorf("mls.credential: CredentialType type not allows")
+	}
+}
+
+func (c *Credential) UnmarshalTLS(data []byte) (int, error) {
+	switch c.CredentialType {
+	case credential_basic:
+		var inner BasicCredential
```

Comment:
You don't need to define a new value here, you can unmarshal directly into the member.

```
read, err = syntax.Unmarshal(data, c.BasicCredential)
```

Reply:
(That's assuming you make the member a pointer.)